## Q1 In this question provide sentence inside val and run it will provide unicode correction and length

In [4]:
with open('hi_100.txt', 'r', encoding='utf-8') as f:
        corpus = f.readlines()
        lines = corpus


In [5]:
consonants = {
    "क": "k", "ख": "kh", "ग": "g", "घ": "gh", "ङ": "N",
    "च": "ch", "छ": "chh", "ज": "j", "झ": "jh", "ञ": "~N",
    "ट": "T", "ठ": "Th", "ड": "D", "ढ": "Dh", "ण": "N",
    "त": "t", "थ": "th", "द": "d", "ध": "dh", "न": "n",
    "प": "p", "फ": "ph", "ब": "b", "भ": "bh", "म": "m",
    "य": "y", "र": "r", "ल": "l", "व": "v", "श": "sh",
    "ष": "Sh", "स": "s", "ह": "h", "ळ": "L", "क्ष": "kSh",
    "ज्ञ": "j~n", "ड़": "R", "य़": "y", "ज़": "z", "ब़": "b", "क़": "q", "ख़": "Kh", "ग़": "G",
    "ड़": "R", "ढ़": "Rh", "फ़": "f", "श़": "sh", "ऋ": "Ri", "ॠ": "Ri", "ॡ": "Li", "ऌ": "Li", "ऴ": "L", "ॐ": "OM",
    "ऽ": "'"
         ""
}
hindi_to_english_transliteration = {
    # Vowels
    "अ": "a", "आ": "aa", "इ": "i", "ई": "ii", "उ": "u",
    "ऊ": "uu", "ऋ": "RRi", "ॠ": "RRi", "ऌ": "LLi", "ॡ": "LLi",
    "ए": "e", "ऐ": "ai", "ओ": "o", "औ": "au", "अं": "am", "अः": "ah",
}
matras = {"ा": "आ", "ि": "इ", "ी": "ई", "ु": "उ", "ू": "ऊ", "ृ": "ऋ", "ॄ": "ॠ", "ॢ": "ऌ", "ॣ": "ॡ", "े": "ए", "ै": "ऐ",
          "ो": "ओ", "ौ": "औ", "ं": "अं", "ः": "अः", "ँ": "अँ"}
reverse_matras = {"अ": "", "आ": "ा", "इ": "ि", "ई": "ी", "उ": "ु", "ऊ": "ू", "ऋ": "ृ", "ॠ": "ॄ", "ऌ": "ॢ", "ॡ": "ॣ",
                  "ए": "े", "ऐ": "ै", "ओ": "ो", "औ": "ौ", "अं": "ं", "अः": "ः", "अँ": "ँ"}
halanta = "्"
halanta_char = []
second_matras = {"अं": "ं", "अः": "ः", "अँ": "ँ"}
reverse_matras_second = {"ं": "अं", "ः": "अः", "ँ": "अँ"}

In [6]:
h_char = consonants.keys()
for x in h_char:
    halanta_char.append(x + halanta)

#### Character Conversion

In [7]:
def character_conversion(line):
    one = []
    for lineIndex in range(len(line)):
        if line[lineIndex] in consonants and lineIndex + 1 < len(line) and line[lineIndex + 1] != '्':
            if line[lineIndex + 1] in matras:
                one.append(line[lineIndex])
                one.append(matras[line[lineIndex + 1]])
                lineIndex += 1
                if lineIndex+1<len(line) and line[lineIndex + 1] in matras:
                    one.append(matras[line[lineIndex + 1]])
                    lineIndex += 1
            elif line[lineIndex + 1] == "़":
                one.append(line[lineIndex] + "़")
                if lineIndex + 2 < len(line) and line[lineIndex + 2] in second_matras:
                    one.append(matras[line[lineIndex + 2]])
                    lineIndex += 2
                else:
                    one.append('अ')
            else:
                one.append(line[lineIndex])
                one.append('अ')
        elif line[lineIndex] in consonants and lineIndex + 1 < len(line) and line[lineIndex + 1] == '्':
            val = line[lineIndex]
            val += "्"
            one.append(val)
            lineIndex += 1
        elif line[lineIndex] in hindi_to_english_transliteration:
            one.append(line[lineIndex])
        # elif line[lineIndex] == " ":
        #     one.append(" ")
        else:
            continue
    return one

In [8]:
def char_to_halanta(val):
    for x in range(len(val)-1):
        if val[x] in consonants:
            val[x] = val[x]+"्"
    return val

In [9]:
val = character_conversion("तुम्हारा नाम क्या हे?")
print(char_to_halanta(val))
print(len(val))

['त', 'उ', 'म्', 'ह', 'आ', 'र', 'आ', 'न', 'आ', 'म', 'अ', 'क्', 'य', 'आ', 'ह', 'ए']
16


## Q2 Finding character and syllables unigram and bigrams

In [10]:
def topFreqUnigram(u_freq,k):
    sorted_unigram_freq = sorted(u_freq.items(), key=lambda x: x[1], reverse=True)
    top_k_unigram = dict(sorted_unigram_freq[:k])
    for items in top_k_unigram:
        print(f"{items}: {top_k_unigram[items]}")
    return top_k_unigram

def topFreqBigram(b_freq,k):
    sorted_bigram_freq = sorted(b_freq.items(), key=lambda x: x[1], reverse=True)
    top_k_bigram = dict(sorted_bigram_freq[:k])
    for items in top_k_bigram:
        print(f"{items}: {top_k_bigram[items]}")
    return top_k_bigram

#### Syllable Formation

In [11]:
def syllable_convert(val):
    a = ""
    one = []
    i = 0
    while i < len(val):
        # if val[i] == " ":
        #     # one.append(" ")
        #     i += 1
        #     continue
        # print(i)
        if val[i] in consonants:
            a = a + val[i]
        elif val[i] in halanta_char:
            a = a + val[i]
        else:
            if a != "":
                if val[i] in reverse_matras:
                    a = a + reverse_matras[val[i]]
                if i + 1 < len(val) and val[i + 1] in second_matras:
                    a = a + second_matras[val[i + 1]]
                    i = i + 1
            else:
                a = val[i]
            one.append(a)
            a = ""
        i += 1
        # else:
        #     one.append(val[i])
    for x in one:
        if x == "":
            one.remove(x)
    return one

#### Finding Frequencies

In [12]:
%%time
unigram_freq = {}
unigram_freq_char = {}
bigram_freq = {}
bigram_freq_char = {}
x = 0
for line in lines:
    val = character_conversion(line)
    freq = syllable_convert(val)
    for it in val:
        if it in unigram_freq:
            unigram_freq_char[it] += 1
        else:
            unigram_freq_char[it] = 1
    for item in freq:
        if item in unigram_freq:
            unigram_freq[item] += 1
        else:
            unigram_freq[item] = 1

    for it1 in range(len(val) - 1):
        bigram = " ".join((val[it1], val[it1 + 1]))
        if bigram in bigram_freq_char:
            bigram_freq_char[bigram] += 1
        else:
            bigram_freq_char[bigram] = 1
    for i in range(len(freq) - 1):
        bigram = " ".join((freq[i], freq[i + 1]))
        if bigram in bigram_freq:
            bigram_freq[bigram] += 1
        else:
            bigram_freq[bigram] = 1
# # top 20 values
print("Syllable Unigram Frequency:")
topFreqUnigram(unigram_freq,20)
print("Syllable Bigram Frequency:")
topFreqBigram(bigram_freq,20)
print("Character Unigram Frequency:")
topFreqUnigram(unigram_freq_char,20)
print("Character Bigram Frequency:")
topFreqBigram(bigram_freq_char,20)

Syllable Unigram Frequency:
र: 1007332
क: 604341
न: 504714
स: 489101
के: 397246
प: 391011
ल: 328365
ने: 325175
त: 303783
का: 294287
म: 293458
ए: 287987
ह: 278528
में: 259063
अ: 253328
ब: 246861
की: 234951
ग: 225107
है: 216182
या: 213712
Syllable Bigram Frequency:
क र: 161771
औ र: 115585
प र: 99975
इ स: 83107
ए क: 59487
लि ए: 54067
के लि: 50098
न हीं: 47399
अ प: 45193
र ने: 44066
त क: 40316
का र: 39821
कि या: 37105
ता है: 36522
ने के: 35282
स के: 34216
न के: 32861
क हा: 32812
र का: 32562
य ह: 31774
Character Unigram Frequency:
अ: 7153322
आ: 2966629
ए: 2303822
क: 2059392
र: 1885114
ई: 1445367
इ: 1425781
न: 1250677
ह: 1130489
स: 1115612
म: 1022283
ओ: 891886
ल: 862535
त: 852908
य: 752198
प: 664385
व: 604494
उ: 586139
द: 546709
ज: 514198
Character Bigram Frequency:
र अ: 1173240
अ क: 921468
अ र: 786891
क अ: 616852
अ ह: 562925
अ न: 551313
स अ: 518739
न अ: 515816
अ म: 467030
क ए: 407129
प अ: 405636
त अ: 353263
अ स: 348242
आ र: 343947
ए अं: 341741
ल अ: 333296
अ त: 332006
न ए: 328953
म अ: 324890

{'र अ': 1173240,
 'अ क': 921468,
 'अ र': 786891,
 'क अ': 616852,
 'अ ह': 562925,
 'अ न': 551313,
 'स अ': 518739,
 'न अ': 515816,
 'अ म': 467030,
 'क ए': 407129,
 'प अ': 405636,
 'त अ': 353263,
 'अ स': 348242,
 'आ र': 343947,
 'ए अं': 341741,
 'ल अ': 333296,
 'अ त': 332006,
 'न ए': 328953,
 'म अ': 324890,
 'क आ': 314321}

## Q4 Finding unigram frequency of token and bigram frequencies of tokens, syllables and characters.

#### Loading File

In [13]:
file_path = 'hi_100.txt'  
try:
    with open(file_path, 'r',encoding='utf-8') as file:
        data = file.read()
        print(data[:1000])

except FileNotFoundError:
    print(f"The file '{file_path}' was not found.")
except Exception as e:
    print(f"An error occurred: {e}")


def countF(tokenizer,data):
    tokens = tokenizer.encode(data, out_type=str)
    unigram_frequencies = {}
    for token in tokens:
        if token in unigram_frequencies:
            unigram_frequencies[token] += 1
        else:
            unigram_frequencies[token] = 1
    bigrams = [(tokens[i], tokens[i + 1]) for i in range(len(tokens) - 1)]
    bigram_frequencies = {}
    for bigram in bigrams:
        if bigram in bigram_frequencies:
            bigram_frequencies[bigram] += 1
        else:
            bigram_frequencies[bigram] = 1
    syllables = []
    characters = []
    for tok in tokens:
        syllables.extend(syllable_convert(tok))
        characters.extend(character_conversion(tok))
    bigram_syllables = [(syllables[i], syllables[i + 1]) for i in range(len(syllables) - 1)]
    bigram_characters = [(characters[i], characters[i + 1]) for i in range(len(characters) - 1)]
    syllable_freqs = Counter(bigram_syllables)
    char_freqs = Counter(bigram_characters)
    return unigram_frequencies, bigram_frequencies, syllable_freqs, char_freqs

आवेदन करने की आखिरी तारीख 31 जनवरी, 2020 है।
इतनी दुआ कर दो हमारे लिए कि जितना प्यार दुनिया ने आपको दिया है, बस उतना ही हमें भी मिल जाए|”
मोदी सरकार के पहले कार्यकाल में भी तीन तलाक को लेकर बिल लाया गया था, हालांकि तब यह राज्यसभा में पास नहीं हो पाया था.
भाजपा के दिवंगत नेता प्रमोद महाजन की बेटी पूनम महाजन को सचिव बनाया गया है.
ऐसी स्थिति में एक न्यायपूर्ण सरकार सार्वजनिक वित्त का इस तरह इस्तेमाल करती है कि संसाधनों का आवंटन, सभी के उपभोग वाले उत्पादों की व्यवहार्यता और समग्र वृहद-आर्थिक प्रबंधन 'निष्पक्षता के रूप में न्याय' को बढ़ाए।
दिलचस्प है कि डीसीएचएल के चेयरमैन टी वेंकटरमन रेड्डी और वाइस चेयरमैन टी विनायक रवि रेड्डी इस बैठक में मौजूद नहीं थे।
इस आम चुनाव में भाजपा नेता सतीश कुमार गौतम को सबसे अधिक 6 लाख 56 हजार 215 वोट प्राप्त हुए.
आयरलैंड टीम के विकेटकीपर बल्लेबाज नियाल ओ'ब्रायन ने अंतर्राष्ट्रीय क्रिकेट से संन्यास का ऐलान कर दिया है। साल 2002 में डेनमार्क के खिलाफ के एकदिवसीय क्रिकेट में डेब्यू करने वाले ब्रायन ने 36 साल की उम्र में संन्यास का ऐलान किया। नियाल ओ'ब्रायन आयरलैंड

#### BPE and Unigram Functions

In [14]:
import sentencepiece as spm
from collections import Counter
import tensorflow as tf
import time

def unigram_tokenizer(file_path, v_size):
    spm.SentencePieceTrainer.Train(input=file_path, model_prefix='uni_model'+str(v_size),model_type='unigram', vocab_size=v_size)
    model_file='uni_model'+str(v_size)+'.model'
    tokenizer_unigram = spm.SentencePieceProcessor('uni_model'+str(v_size)+'.model')
    return tokenizer_unigram

def bpe_tokenizer(file_path, v_size):
    spm.SentencePieceTrainer.train(input=file_path, model_prefix='bpe_model'+str(v_size), model_type='bpe', vocab_size=v_size)
    bpe_tknizer = spm.SentencePieceProcessor('bpe_model'+str(v_size)+'.model')
    return bpe_tknizer


#### Saving Files

In [18]:
#Function for writing files
def model_save(model_name,v_size,arr):
    uni,bi,syll,char = arr 
    s_uni = sorted(uni.items(), key=lambda x: x[1], reverse=True)
    file_path = ""
    file_name = model_name+'_unigram_tokens_'+str(v_size)+'.txt'
    with open(file_path+file_name, 'w', encoding='utf-8') as f:
        for word, count in s_uni:
            f.write(f"{word}: {count}\n")

    s_bi = sorted(bi.items(), key=lambda x: x[1], reverse=True)
    file_name = model_name + "_bigram_tokens_" +str(v_size)+ ".txt"
    with open(file_path+file_name, 'w', encoding='utf-8') as f:
        for (word1, word2), count in s_bi:
            f.write(f"{word1} {word2}: {count}\n")

    s_syll = sorted(syll.items(), key=lambda x: x[1], reverse=True)
    file_name = model_name + "_bi_syllables_" +str(v_size)+ ".txt"
    with open(file_path+file_name, 'w', encoding='utf-8') as f:
        for (word1, word2), count in s_syll:
            f.write(f"{word1} {word2}: {count}\n")
    s_char = sorted(char.items(), key=lambda x: x[1], reverse=True)
    file_name = model_name + "_bi_char_" +str(v_size)+ ".txt"
    with open(file_path+file_name, 'w', encoding='utf-8') as f:
        for (word1, word2), count in s_char:
            f.write(f"{word1} {word2}: {count}\n")

#### Unigram_1000 vocab

In [19]:
unigram_1000_tokenizer = unigram_tokenizer('hi_100.txt', 1000)
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF(unigram_1000_tokenizer, data)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="unigram",v_size=1000,arr=arr)
print("uni")
topFreqUnigram(unigram_freqs,20)
print("bi")
topFreqUnigram(bigram_freqs,20)
print("syllable")
topFreqUnigram(syllable_freqs,20)
print("char")
topFreqUnigram(char_freqs,20)

uni
र: 431118
▁के: 328920
न: 281629
।: 276966
ा: 256523
ल: 244768
▁में: 240292
▁: 233755
▁है: 215655
ी: 211097
▁की: 199499
,: 191476
म: 185049
े: 184611
▁को: 159954
▁से: 154736
क: 154277
त: 145275
ो: 137056
ि: 134016
bi
('▁है', '।'): 89363
('▁के', '▁लिए'): 43425
('▁हैं', '।'): 33632
('▁है', '.'): 25838
('▁है', '▁कि'): 25125
('ों', '▁के'): 21724
('र', 'ा'): 19993
('ों', '▁में'): 19869
('▁है', ','): 19551
('▁के', '▁साथ'): 18346
('र', 'ी'): 17879
('ते', '▁हैं'): 17736
('ा', 'र'): 17496
('ने', '▁के'): 17090
('ों', '▁को'): 16905
('ता', '▁है'): 16224
('▁कहा', '▁कि'): 16099
('।', '▁इस'): 14551
('▁के', '▁बाद'): 14451
('ों', '▁की'): 14109
syllable
('▁', 'क'): 1164117
('क', '▁'): 1059456
('ं', '▁'): 752803
('▁', 'ह'): 522370
('▁', 'म'): 464845
('▁', 'स'): 452985
('▁', '▁'): 418607
('न', '▁'): 393052
('स', '▁'): 325851
('त', '▁'): 284825
('र', '▁'): 279269
('।', '▁'): 272992
('म', 'ं'): 264559
('▁', 'अ'): 245641
('▁', 'प'): 241973
('▁', 'न'): 236433
('▁', 'द'): 209848
('य', '▁'): 205941
('ल', '▁'

{('क', 'ए'): 400441,
 ('आ', 'क'): 397593,
 ('न', 'ए'): 323222,
 ('क', 'आ'): 300427,
 ('ह', 'ऐ'): 296118,
 ('क', 'अ'): 292726,
 ('य', 'आ'): 288660,
 ('ए', 'क'): 274585,
 ('अ', 'ह'): 272906,
 ('ए', 'अं'): 270356,
 ('र', 'अ'): 267824,
 ('म', 'ए'): 267626,
 ('अ', 'क'): 263891,
 ('प', 'अ'): 248206,
 ('क', 'ई'): 232707,
 ('अ', 'न'): 208993,
 ('स', 'ए'): 207418,
 ('आ', 'ह'): 206676,
 ('क', 'ओ'): 198975,
 ('स', 'अ'): 196480}

#### Unigram_2000

In [20]:
unigram_2000_tokenizer = unigram_tokenizer('hi_100.txt', 2000)
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF(unigram_2000_tokenizer, data)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="unigram",v_size=2000,arr=arr)
topFreqUnigram(unigram_freqs,20)
topFreqUnigram(bigram_freqs,20)
topFreqUnigram(syllable_freqs,20)
topFreqUnigram(char_freqs,20)

▁के: 324723
।: 273666
▁में: 240292
▁है: 215632
▁की: 196738
,: 190026
ी: 173237
▁को: 154736
▁से: 151426
ल: 146463
र: 145878
न: 142629
▁का: 131131
▁: 126257
.: 121201
ा: 117568
▁और: 115376
क: 110608
ों: 109310
▁ने: 107299
('▁है', '।'): 89363
('▁के', '▁लिए'): 43425
('▁हैं', '।'): 33632
('▁है', '.'): 25841
('▁है', '▁कि'): 25121
('▁है', ','): 19551
('ों', '▁के'): 19271
('▁के', '▁साथ'): 18346
('ों', '▁में'): 17781
('▁कहा', '▁कि'): 16097
('ों', '▁को'): 14835
('▁के', '▁बाद'): 14451
('▁था', '।'): 12626
('ों', '▁की'): 12609
('▁रहा', '▁है'): 12021
('ने', '▁के'): 11985
('▁ने', '▁कहा'): 11830
('▁हैं', '.'): 11093
('▁गया', '▁है'): 11051
('।', '▁इस'): 10846
('▁', 'क'): 1153274
('क', '▁'): 1042132
('ं', '▁'): 738867
('▁', 'ह'): 524734
('▁', 'म'): 469974
('▁', 'स'): 459403
('▁', '▁'): 418607
('न', '▁'): 375091
('स', '▁'): 330014
('र', '▁'): 314242
('।', '▁'): 272992
('म', 'ं'): 269732
('त', '▁'): 269345
('▁', 'प'): 263727
('▁', 'अ'): 244256
('▁', 'न'): 230018
('▁', 'द'): 220728
('य', '▁'): 213129
('▁',

{('क', 'ए'): 399726,
 ('आ', 'क'): 383301,
 ('क', 'अ'): 336247,
 ('र', 'अ'): 333141,
 ('न', 'ए'): 321740,
 ('अ', 'क'): 316580,
 ('अ', 'ह'): 310675,
 ('क', 'आ'): 307132,
 ('ह', 'ऐ'): 296084,
 ('य', 'आ'): 291860,
 ('ए', 'अं'): 291735,
 ('म', 'ए'): 290675,
 ('प', 'अ'): 287430,
 ('ए', 'क'): 272411,
 ('अ', 'र'): 253669,
 ('स', 'अ'): 246782,
 ('अ', 'न'): 238023,
 ('क', 'ई'): 231881,
 ('आ', 'ह'): 215208,
 ('स', 'ए'): 212515}

#### BPE_1000

In [22]:
bpe_1000_tokenizer = bpe_tokenizer('hi_100.txt', 1000)
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF(bpe_1000_tokenizer, data)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="bpe",v_size=1000,arr=arr)
topFreqUnigram(unigram_freqs,20)
topFreqUnigram(bigram_freqs,20)
topFreqUnigram(syllable_freqs,20)
topFreqUnigram(char_freqs,20)

▁के: 328806
।: 270047
▁में: 240292
▁है: 215519
▁की: 199499
,: 186960
▁को: 162462
ल: 158685
न: 156906
▁से: 152622
क: 144398
त: 140020
▁: 133468
▁का: 123398
म: 123257
र: 121598
.: 116875
▁और: 115365
े: 112549
▁ने: 110224
('▁है', '।'): 89363
('▁के', '▁लिए'): 43425
('▁हैं', '।'): 33632
('▁है', '.'): 25579
('▁है', '▁कि'): 25130
('▁है', ','): 19551
('▁के', '▁साथ'): 18346
('ों', '▁के'): 17977
('ने', '▁के'): 16651
('▁कहा', '▁कि'): 16100
('ों', '▁में'): 16067
('ते', '▁हैं'): 15852
('▁के', '▁बाद'): 14451
('ों', '▁को'): 14254
('ता', '▁है'): 13198
('।', '▁इस'): 12725
('▁था', '।'): 12626
('▁रहा', '▁है'): 12021
('▁ने', '▁कहा'): 11824
('ों', '▁की'): 11499
('▁', 'क'): 1146933
('क', '▁'): 1070351
('ं', '▁'): 751497
('▁', 'ह'): 536002
('▁', 'स'): 454872
('▁', 'म'): 429313
('▁', '▁'): 418607
('न', '▁'): 406735
('र', '▁'): 347487
('स', '▁'): 342313
('त', '▁'): 303380
('।', '▁'): 272992
('म', 'ं'): 272813
('▁', 'अ'): 245641
('▁', 'न'): 233158
('▁', 'प'): 220766
('य', '▁'): 216625
('ल', '▁'): 207526
('▁', '

{('क', 'ए'): 393586,
 ('आ', 'क'): 386568,
 ('न', 'ए'): 325957,
 ('क', 'अ'): 304719,
 ('ह', 'ऐ'): 296118,
 ('य', 'आ'): 288185,
 ('क', 'आ'): 287318,
 ('म', 'ए'): 274562,
 ('ए', 'क'): 273311,
 ('ए', 'अं'): 271900,
 ('अ', 'क'): 258782,
 ('अ', 'ह'): 242105,
 ('क', 'ई'): 234396,
 ('र', 'अ'): 233647,
 ('प', 'अ'): 231507,
 ('स', 'ए'): 212670,
 ('अ', 'न'): 210328,
 ('क', 'ओ'): 202105,
 ('आ', 'ह'): 200429,
 ('ई', 'क'): 194279}

#### BPE_2000

In [23]:
bpe_2000_tokenizer = bpe_tokenizer('hi_100.txt', 2000)
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF(bpe_2000_tokenizer, data)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="bpe",v_size=2000,arr=arr)
topFreqUnigram(unigram_freqs,20)
topFreqUnigram(bigram_freqs,20)
topFreqUnigram(syllable_freqs,20)
topFreqUnigram(char_freqs,20)

▁के: 326040
।: 270047
▁में: 240292
▁है: 215519
▁की: 197474
,: 185205
▁को: 156151
▁से: 147346
▁का: 123398
.: 116875
▁और: 115365
▁ने: 108464
▁पर: 100405
क: 90482
▁कि: 89667
न: 82236
▁हैं: 80599
ों: 77433
ल: 73155
म: 71367
('▁है', '।'): 89363
('▁के', '▁लिए'): 43425
('▁हैं', '।'): 33632
('▁है', '.'): 25579
('▁है', '▁कि'): 25108
('▁है', ','): 19551
('▁के', '▁साथ'): 18346
('▁कहा', '▁कि'): 16096
('▁के', '▁बाद'): 14451
('ों', '▁के'): 13475
('ों', '▁में'): 13042
('▁था', '।'): 12626
('▁रहा', '▁है'): 12021
('▁ने', '▁कहा'): 11824
('ने', '▁के'): 11590
('ता', '▁है'): 11120
('▁गया', '▁है'): 11051
('▁हैं', '.'): 10997
('।', '▁इस'): 10693
('▁रहे', '▁हैं'): 10180
('▁', 'क'): 1157614
('क', '▁'): 1055357
('ं', '▁'): 743692
('▁', 'ह'): 524524
('▁', 'म'): 471201
('▁', 'स'): 457869
('▁', '▁'): 418607
('न', '▁'): 389912
('स', '▁'): 342860
('र', '▁'): 339484
('त', '▁'): 291970
('।', '▁'): 272992
('म', 'ं'): 271581
('▁', 'प'): 253433
('▁', 'अ'): 245641
('▁', 'न'): 233909
('▁', 'द'): 217402
('य', '▁'): 216467
('

{('क', 'ए'): 402872,
 ('आ', 'क'): 377294,
 ('क', 'अ'): 330433,
 ('अ', 'क'): 326235,
 ('न', 'ए'): 325957,
 ('अ', 'ह'): 304979,
 ('क', 'आ'): 304314,
 ('ह', 'ऐ'): 296118,
 ('ए', 'अं'): 294293,
 ('र', 'अ'): 293990,
 ('य', 'आ'): 291608,
 ('म', 'ए'): 285742,
 ('प', 'अ'): 284357,
 ('ए', 'क'): 275062,
 ('अ', 'र'): 252010,
 ('स', 'अ'): 246606,
 ('अ', 'न'): 245780,
 ('क', 'ई'): 236594,
 ('स', 'ए'): 212670,
 ('आ', 'ह'): 209328}

#### BERT 1000 & 2000

In [24]:
from transformers import BertTokenizer, BertModel
import tensorflow as tf
from collections import Counter
import time
def mBert():
    bertToken = BertTokenizer.from_pretrained('bert-base-multilingual-cased')
    return bertToken

def countF1(tokens):
    u_freq = Counter(tokens)
    bigrams = [(tokens[i], tokens[i + 1]) for i in range(len(tokens) - 1)]
    bigram_frequencies = {}
    for bigram in bigrams:
        if bigram in bigram_frequencies:
            bigram_frequencies[bigram] += 1
        else:
            bigram_frequencies[bigram] = 1
    syllables = []
    char = []
    for tkn in tokens:
        syllables.extend(syllable_convert(tkn))
        char.extend(character_conversion(tkn))
    bi_syllables = [(syllables[i], syllables[i + 1]) for i in range(len(syllables) - 1)]
    bi_characters = [(char[i], char[i + 1]) for i in range(len(char) - 1)]
    syllable_freqs = Counter(bi_syllables)
    char_freqs = Counter(bi_characters)
    return u_freq, bigram_frequencies, syllable_freqs, char_freqs
    

In [25]:
mBertTokenizer = mBert()
mB_encoding = mBertTokenizer.encode_plus(data, max_length=1000, truncation=True)
mBert_tokens = mBertTokenizer.convert_ids_to_tokens(mB_encoding['input_ids'])
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF1(mBert_tokens)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="mBert",v_size=1000,arr=arr)
topFreqUnigram(unigram_freqs,30)
topFreqUnigram(bigram_freqs,30)
topFreqUnigram(syllable_freqs,30)
topFreqUnigram(char_freqs,30)


के: 22
में: 20
।: 18
है: 15
ब: 15
##प: 14
आ: 13
,: 13
को: 13
##ी: 12
##र: 12
और: 12
##ा: 11
ख: 11
##न: 10
की: 9
##स: 8
म: 8
स: 8
का: 8
हैं: 8
कर: 7
कि: 7
ज: 7
##द: 7
##्: 7
व: 7
##े: 7
##ोज: 7
प: 6
('आ', '##प'): 7
('ख', '##ोज'): 7
('है', '।'): 6
('है', 'कि'): 4
('हैं', ','): 4
('##ोज', 'शब्द'): 4
('##ों', 'की'): 3
('##्', '##प'): 3
('को', 'ब'): 3
('##रल', '##ैंड'): 3
('वि', '##के'): 3
('ब', '##्रा'): 3
('##्रा', '##यन'): 3
('के', 'लिए'): 3
('शब्द', '##ों'): 3
('ज', '##ित'): 2
('##ित', '##ना'): 2
('##प', '##को'): 2
('दिया', 'है'): 2
('है', ','): 2
('उ', '##तन'): 2
('##तन', '##ा'): 2
('##ा', 'ही'): 2
('जा', '##ए'): 2
('में', 'भी'): 2
('ब', '##िल'): 2
('भ', '##ाज'): 2
('##ाज', '##पा'): 2
('ने', '##ता'): 2
('म', '##हा'): 2
('#', '#'): 549
('#', '्'): 51
('#', 'ि'): 37
('क', '#'): 34
('्', '#'): 34
('#', 'क'): 31
('#', 'े'): 24
('#', 'ो'): 20
('म', 'ं'): 20
('ं', '#'): 20
('ि', '#'): 19
('आ', '#'): 18
('#', 'ा'): 18
('ं', 'क'): 16
('#', 'ी'): 13
('#', 'न'): 12
('े', '#'): 12
('न', '#'): 11


{('आ', 'क'): 31,
 ('क', 'ए'): 30,
 ('र', 'अ'): 25,
 ('अ', 'क'): 24,
 ('क', 'अ'): 23,
 ('ह', 'ऐ'): 23,
 ('य', 'आ'): 22,
 ('अ', 'न'): 21,
 ('अ', 'र'): 20,
 ('म', 'ए'): 20,
 ('ए', 'अं'): 20,
 ('न', 'ए'): 16,
 ('क', 'ओ'): 16,
 ('अ', 'ह'): 16,
 ('आ', 'ह'): 15,
 ('क', 'इ'): 14,
 ('क', 'आ'): 14,
 ('प', 'अ'): 14,
 ('न', 'आ'): 13,
 ('स', 'अ'): 12,
 ('इ', 'क'): 12,
 ('अ', 'त'): 12,
 ('क', 'ई'): 11,
 ('आ', 'य'): 11,
 ('ल', 'आ'): 10,
 ('र', 'आ'): 10,
 ('ए', 'क'): 9,
 ('आ', 'र'): 9,
 ('ज', 'अ'): 9,
 ('ह', 'अ'): 9}

In [26]:
mBertTokenizer_2000 = mBert()
mB_encoding_2000 = mBertTokenizer.encode_plus(data, max_length=2000, truncation=True)
mBert_tokens_2000 = mBertTokenizer_2000.convert_ids_to_tokens(mB_encoding_2000['input_ids'])
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF1(mBert_tokens)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="mBert",v_size=2000,arr=arr)
topFreqUnigram(unigram_freqs,30)
topFreqUnigram(bigram_freqs,30)
topFreqUnigram(syllable_freqs,30)
topFreqUnigram(char_freqs,30)

के: 22
में: 20
।: 18
है: 15
ब: 15
##प: 14
आ: 13
,: 13
को: 13
##ी: 12
##र: 12
और: 12
##ा: 11
ख: 11
##न: 10
की: 9
##स: 8
म: 8
स: 8
का: 8
हैं: 8
कर: 7
कि: 7
ज: 7
##द: 7
##्: 7
व: 7
##े: 7
##ोज: 7
प: 6
('आ', '##प'): 7
('ख', '##ोज'): 7
('है', '।'): 6
('है', 'कि'): 4
('हैं', ','): 4
('##ोज', 'शब्द'): 4
('##ों', 'की'): 3
('##्', '##प'): 3
('को', 'ब'): 3
('##रल', '##ैंड'): 3
('वि', '##के'): 3
('ब', '##्रा'): 3
('##्रा', '##यन'): 3
('के', 'लिए'): 3
('शब्द', '##ों'): 3
('ज', '##ित'): 2
('##ित', '##ना'): 2
('##प', '##को'): 2
('दिया', 'है'): 2
('है', ','): 2
('उ', '##तन'): 2
('##तन', '##ा'): 2
('##ा', 'ही'): 2
('जा', '##ए'): 2
('में', 'भी'): 2
('ब', '##िल'): 2
('भ', '##ाज'): 2
('##ाज', '##पा'): 2
('ने', '##ता'): 2
('म', '##हा'): 2
('#', '#'): 549
('#', '्'): 51
('#', 'ि'): 37
('क', '#'): 34
('्', '#'): 34
('#', 'क'): 31
('#', 'े'): 24
('#', 'ो'): 20
('म', 'ं'): 20
('ं', '#'): 20
('ि', '#'): 19
('आ', '#'): 18
('#', 'ा'): 18
('ं', 'क'): 16
('#', 'ी'): 13
('#', 'न'): 12
('े', '#'): 12
('न', '#'): 11


{('आ', 'क'): 31,
 ('क', 'ए'): 30,
 ('र', 'अ'): 25,
 ('अ', 'क'): 24,
 ('क', 'अ'): 23,
 ('ह', 'ऐ'): 23,
 ('य', 'आ'): 22,
 ('अ', 'न'): 21,
 ('अ', 'र'): 20,
 ('म', 'ए'): 20,
 ('ए', 'अं'): 20,
 ('न', 'ए'): 16,
 ('क', 'ओ'): 16,
 ('अ', 'ह'): 16,
 ('आ', 'ह'): 15,
 ('क', 'इ'): 14,
 ('क', 'आ'): 14,
 ('प', 'अ'): 14,
 ('न', 'आ'): 13,
 ('स', 'अ'): 12,
 ('इ', 'क'): 12,
 ('अ', 'त'): 12,
 ('क', 'ई'): 11,
 ('आ', 'य'): 11,
 ('ल', 'आ'): 10,
 ('र', 'आ'): 10,
 ('ए', 'क'): 9,
 ('आ', 'र'): 9,
 ('ज', 'अ'): 9,
 ('ह', 'अ'): 9}

#### INDI_BERT 1000_2000

In [27]:
import transformers
import collections
def tokenizer_IndicBert():
    tokenizer_indiBert = transformers.AutoTokenizer.from_pretrained("ai4bharat/indic-bert",lang='hi')
    return tokenizer_indiBert

In [28]:
indi_tokenizer = tokenizer_IndicBert()
indi_token = indi_tokenizer(data,max_length=1000, truncation=True)
indi_tokens = indi_tokenizer.convert_ids_to_tokens(indi_token['input_ids'])
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF1(mBert_tokens)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="IndicBert",v_size=1000,arr=arr)
topFreqUnigram(unigram_freqs,30)
topFreqUnigram(bigram_freqs,30)
topFreqUnigram(syllable_freqs,30)
topFreqUnigram(char_freqs,30)

के: 22
में: 20
।: 18
है: 15
ब: 15
##प: 14
आ: 13
,: 13
को: 13
##ी: 12
##र: 12
और: 12
##ा: 11
ख: 11
##न: 10
की: 9
##स: 8
म: 8
स: 8
का: 8
हैं: 8
कर: 7
कि: 7
ज: 7
##द: 7
##्: 7
व: 7
##े: 7
##ोज: 7
प: 6
('आ', '##प'): 7
('ख', '##ोज'): 7
('है', '।'): 6
('है', 'कि'): 4
('हैं', ','): 4
('##ोज', 'शब्द'): 4
('##ों', 'की'): 3
('##्', '##प'): 3
('को', 'ब'): 3
('##रल', '##ैंड'): 3
('वि', '##के'): 3
('ब', '##्रा'): 3
('##्रा', '##यन'): 3
('के', 'लिए'): 3
('शब्द', '##ों'): 3
('ज', '##ित'): 2
('##ित', '##ना'): 2
('##प', '##को'): 2
('दिया', 'है'): 2
('है', ','): 2
('उ', '##तन'): 2
('##तन', '##ा'): 2
('##ा', 'ही'): 2
('जा', '##ए'): 2
('में', 'भी'): 2
('ब', '##िल'): 2
('भ', '##ाज'): 2
('##ाज', '##पा'): 2
('ने', '##ता'): 2
('म', '##हा'): 2
('#', '#'): 549
('#', '्'): 51
('#', 'ि'): 37
('क', '#'): 34
('्', '#'): 34
('#', 'क'): 31
('#', 'े'): 24
('#', 'ो'): 20
('म', 'ं'): 20
('ं', '#'): 20
('ि', '#'): 19
('आ', '#'): 18
('#', 'ा'): 18
('ं', 'क'): 16
('#', 'ी'): 13
('#', 'न'): 12
('े', '#'): 12
('न', '#'): 11


{('आ', 'क'): 31,
 ('क', 'ए'): 30,
 ('र', 'अ'): 25,
 ('अ', 'क'): 24,
 ('क', 'अ'): 23,
 ('ह', 'ऐ'): 23,
 ('य', 'आ'): 22,
 ('अ', 'न'): 21,
 ('अ', 'र'): 20,
 ('म', 'ए'): 20,
 ('ए', 'अं'): 20,
 ('न', 'ए'): 16,
 ('क', 'ओ'): 16,
 ('अ', 'ह'): 16,
 ('आ', 'ह'): 15,
 ('क', 'इ'): 14,
 ('क', 'आ'): 14,
 ('प', 'अ'): 14,
 ('न', 'आ'): 13,
 ('स', 'अ'): 12,
 ('इ', 'क'): 12,
 ('अ', 'त'): 12,
 ('क', 'ई'): 11,
 ('आ', 'य'): 11,
 ('ल', 'आ'): 10,
 ('र', 'आ'): 10,
 ('ए', 'क'): 9,
 ('आ', 'र'): 9,
 ('ज', 'अ'): 9,
 ('ह', 'अ'): 9}

In [29]:
indi_tokenizer_2000 = tokenizer_IndicBert()
indi_token_2000 = indi_tokenizer_2000(data,max_length=2000, truncation=True)
indi_tokens_2000 = indi_tokenizer_2000.convert_ids_to_tokens(indi_token_2000['input_ids'])
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF1(mBert_tokens)
arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
model_save(model_name="IndicBert",v_size=2000,arr=arr)
topFreqUnigram(unigram_freqs,30)
topFreqUnigram(bigram_freqs,30)
topFreqUnigram(syllable_freqs,30)
topFreqUnigram(char_freqs,30)

के: 22
में: 20
।: 18
है: 15
ब: 15
##प: 14
आ: 13
,: 13
को: 13
##ी: 12
##र: 12
और: 12
##ा: 11
ख: 11
##न: 10
की: 9
##स: 8
म: 8
स: 8
का: 8
हैं: 8
कर: 7
कि: 7
ज: 7
##द: 7
##्: 7
व: 7
##े: 7
##ोज: 7
प: 6
('आ', '##प'): 7
('ख', '##ोज'): 7
('है', '।'): 6
('है', 'कि'): 4
('हैं', ','): 4
('##ोज', 'शब्द'): 4
('##ों', 'की'): 3
('##्', '##प'): 3
('को', 'ब'): 3
('##रल', '##ैंड'): 3
('वि', '##के'): 3
('ब', '##्रा'): 3
('##्रा', '##यन'): 3
('के', 'लिए'): 3
('शब्द', '##ों'): 3
('ज', '##ित'): 2
('##ित', '##ना'): 2
('##प', '##को'): 2
('दिया', 'है'): 2
('है', ','): 2
('उ', '##तन'): 2
('##तन', '##ा'): 2
('##ा', 'ही'): 2
('जा', '##ए'): 2
('में', 'भी'): 2
('ब', '##िल'): 2
('भ', '##ाज'): 2
('##ाज', '##पा'): 2
('ने', '##ता'): 2
('म', '##हा'): 2
('#', '#'): 549
('#', '्'): 51
('#', 'ि'): 37
('क', '#'): 34
('्', '#'): 34
('#', 'क'): 31
('#', 'े'): 24
('#', 'ो'): 20
('म', 'ं'): 20
('ं', '#'): 20
('ि', '#'): 19
('आ', '#'): 18
('#', 'ा'): 18
('ं', 'क'): 16
('#', 'ी'): 13
('#', 'न'): 12
('े', '#'): 12
('न', '#'): 11


{('आ', 'क'): 31,
 ('क', 'ए'): 30,
 ('र', 'अ'): 25,
 ('अ', 'क'): 24,
 ('क', 'अ'): 23,
 ('ह', 'ऐ'): 23,
 ('य', 'आ'): 22,
 ('अ', 'न'): 21,
 ('अ', 'र'): 20,
 ('म', 'ए'): 20,
 ('ए', 'अं'): 20,
 ('न', 'ए'): 16,
 ('क', 'ओ'): 16,
 ('अ', 'ह'): 16,
 ('आ', 'ह'): 15,
 ('क', 'इ'): 14,
 ('क', 'आ'): 14,
 ('प', 'अ'): 14,
 ('न', 'आ'): 13,
 ('स', 'अ'): 12,
 ('इ', 'क'): 12,
 ('अ', 'त'): 12,
 ('क', 'ई'): 11,
 ('आ', 'य'): 11,
 ('ल', 'आ'): 10,
 ('र', 'आ'): 10,
 ('ए', 'क'): 9,
 ('आ', 'र'): 9,
 ('ज', 'अ'): 9,
 ('ह', 'अ'): 9}

#### WhiteSpace

In [30]:
strip_data = data.strip()
if not strip_data:
    white_space_token=[]
else :
    white_space_token = data.split()
unigram_freqs, bigram_freqs, syllable_freqs, char_freqs = countF1(white_space_token)
freq_arr = [unigram_freqs, bigram_freqs, syllable_freqs, char_freqs]
topFreqUnigram(unigram_freqs,30)
topFreqUnigram(bigram_freqs,30)
topFreqUnigram(syllable_freqs,30)
topFreqUnigram(char_freqs,30)

के: 316314
में: 237426
की: 189769
को: 145266
से: 136979
और: 114563
का: 109743
ने: 102378
पर: 88560
है।: 88260
कि: 76741
है: 72656
भी: 63920
एक: 49279
लिए: 47955
इस: 47558
कर: 44563
नहीं: 44168
ही: 40104
तो: 33469
हैं।: 33204
हो: 32014
यह: 30837
करने: 29399
किया: 28818
साथ: 25201
है.: 24954
कहा: 24141
हैं: 23759
बाद: 22786
('के', 'लिए'): 42448
('है', 'कि'): 24816
('के', 'साथ'): 17075
('कहा', 'कि'): 15922
('के', 'बाद'): 14163
('है', 'और'): 10096
('ने', 'कहा'): 9098
('करने', 'के'): 8826
('बताया', 'कि'): 6392
('को', 'लेकर'): 5977
('गया', 'है।'): 5678
('.', '.'): 5638
('रहा', 'है।'): 5483
('के', 'खिलाफ'): 5312
('के', 'दौरान'): 5162
('के', 'बीच'): 5117
('बारे', 'में'): 5059
('करते', 'हुए'): 4817
('रहे', 'हैं।'): 4701
('में', 'भी'): 4682
('कर', 'रहे'): 4637
('जा', 'रहा'): 4633
('हैं', 'और'): 4581
('उन्होंने', 'कहा'): 4578
('रही', 'है।'): 4560
('किया', 'गया'): 4522
('में', 'एक'): 4442
('ने', 'बताया'): 4316
('करने', 'की'): 4280
('साथ', 'ही'): 4259
('म', 'ं'): 271255
('ं', 'क'): 184372
('ह', 'ं'

{('र', 'अ'): 501180,
 ('अ', 'क'): 429218,
 ('क', 'अ'): 422675,
 ('अ', 'र'): 408662,
 ('क', 'ए'): 407129,
 ('अ', 'ह'): 374151,
 ('प', 'अ'): 358976,
 ('अ', 'न'): 358475,
 ('स', 'अ'): 350066,
 ('ए', 'अं'): 341741,
 ('आ', 'क'): 336575,
 ('न', 'ए'): 328953,
 ('क', 'आ'): 314321,
 ('य', 'आ'): 297771,
 ('ह', 'ऐ'): 297199,
 ('म', 'ए'): 296594,
 ('ए', 'क'): 270785,
 ('अ', 'म'): 250370,
 ('अ', 'त'): 243052,
 ('क', 'ई'): 236906,
 ('आ', 'र'): 230627,
 ('न', 'अ'): 226336,
 ('आ', 'ह'): 224131,
 ('स', 'ए'): 216056,
 ('ओ', 'अं'): 212156,
 ('अं', 'क'): 207159,
 ('क', 'ओ'): 204712,
 ('क', 'इ'): 201166,
 ('आ', 'न'): 200173,
 ('र', 'आ'): 199968}

## Q5 Calculating precision, recall and F1 score using answers of Q3 and outputs of tokenizers.

In [31]:
def precision(true_tokens,predicted_tokens):
    true_positives = len(set(predicted_tokens).intersection(set(true_tokens)))
    false_positives = len(set(predicted_tokens).difference(set(true_tokens)))
    return true_positives/(true_positives+false_positives)

def recall(true_tokens,predicted_tokens):
    true_positives = len(set(predicted_tokens).intersection(set(true_tokens)))
    false_negatives = len(set(true_tokens).difference(set(predicted_tokens)))
    return true_positives/(true_positives+false_negatives)

def f1_score(true_tokens,predicted_tokens):
    p = precision(true_tokens,predicted_tokens)
    r = recall(true_tokens,predicted_tokens)
    return 2*p*r/(p+r)

In [32]:
txt = ""
file = 'CS689.txt' 

with open(file, 'r', encoding='utf-8') as f:
    for line in f:
        if line !="\n":
            temp = line.strip()
            txt += temp[3:]
print(txt)

उसकी हालत में सुधार हो रहा है।अज नित नए मीडिया संसथान खुल रहे है, जिसकी फीस हजारों रुपये है। प्रोफेसनल कोर्स होने के चलते छात्र इसमे प्रवेश लेते है। लाखों रुपये खर्च कर पढ़ाई/ प्रशिक्षण प्राप्त करते है। लेकिन उनके सपने तब टूट जाते है जब मीडिया संस्थानों में पहले से जमे दिग्गज ही उनका आर्थिक और मानसिक शोषण शुरू कर देते है।यह आरोप उनपर अंपायर जॉन वार्ड तथा शॉन क्रेग ने लगाए थे।फिर धीरे से गांड को खोल के अपना सुपाड़ा अन्दर कर दिया. सिखा को बड़ा दर्द हुआ और वो कराह उठी.मैंने उसके शोल्डर के ऊपर किस कर दी. और लंड को ऐसे उतना ही अन्दर रहने दिया.ई-गवर्नेंस के तहत मोबाइल फोन जैसे उपकरणों से भी सरकारी सेवाओं तक पहुंचा जा सकता है।बुधवार को शिमला का तापमान 1.4 डिग्री सेल्सियस दर्ज किया गया, जबकि पर्यटन स्थल मनाली में यह शून्य से 2.6 डिग्री सेल्सियस नीचे दर्ज किया गया।उनका स्वास्थ्य गंभीर रूप से प्रभावित था।अभी लीगली बटवारा नहीं हुआ है। सहमति से तलाक़ होने पर मैं अपने पति से अपने लिए या मुझे ना भी मिले तो अपने बेटे के लिए उक्त संपत्ति में हिस्सा माँग सकती हूँ। अगर हाँ तो उस का निर्धारण किस प्रकार से 

In [33]:
file_path = 'CS689_answer.txt'
with open(file_path, 'r', encoding='utf-8') as file:
    tokens = file.read().split(',')
true_tokens = [word.strip() for word in tokens]
true_tokens = [word.replace('\n', '') for word in true_tokens]
print("All true tokens:", true_tokens)

All true tokens: ['उसकी', 'हालत में', 'सुधार', 'हो रहा है।अज', 'नित', 'नए', 'मीडिया संसथान', 'खुल', 'रहे है', 'जिसकी', 'फीस', 'हजारों रुपये', 'है। प्रोफेसनल कोर्स', 'होने के', 'चलते', 'छात्र', 'इसमे', 'प्रवेश', 'लेते है।', 'लाखों रुपये', 'खर्च कर', 'पढ़ाई/ प्रशिक्षण', 'प्राप्त', 'करते है।', 'लेकिन', 'उनके', 'सपने', 'तब', 'टूट', 'जाते है', 'जब', 'मीडिया संस्थानों में', 'पहले से', 'जमे', 'दिग्गज ही', 'उनका', 'आर्थिक', 'और', 'मानसिक', 'शोषण', 'शुरू', 'कर', 'देते है।यह', 'आरोप', 'उनपर', 'अंपायर जॉन वार्ड', 'तथा', 'शॉन क्रेग ने', 'लगाए थे।फिर', 'धीरे से', 'गांड को', 'खोल के', 'अपना', 'सुपाड़ा', 'अन्दर', 'कर दिया.', 'सिखा को', 'बड़ा', 'दर्द', 'हुआ', 'और', 'वो', 'कराह', 'उठी.', 'मैंने', 'उसके', 'शोल्डर के', 'ऊपर', 'किस', 'कर दी.', 'और', 'लंड को', 'ऐसे', 'उतना ही', 'अन्दर', 'रहने दिया.ई-गवर्नेंस के', 'तहत', 'मोबाइल फोन', 'जैसे', 'उपकरणों', 'से भी', 'सरकारी सेवाओं', 'तक', 'पहुंचा', 'जा सकता है।बुधवार को', 'शिमला का', 'तापमान', '1.4 डिग्री सेल्सियस', 'दर्ज', 'किया गया', 'जबकि', 'पर्यटन स्थल', 'मना

#### Unigram_1000 

In [34]:
import re
pred_tokens = unigram_1000_tokenizer.encode(txt, out_type=str)
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]
# Precision
precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.22
Recall: 0.25
F1-Score: 0.23


#### Unigram_2000

In [43]:
pred_tokens = unigram_2000_tokenizer.encode(txt, out_type=str)
# Precision
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]
precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.24
Recall: 0.32
F1-Score: 0.27


#### BPE_1000

In [36]:
pred_tokens = bpe_1000_tokenizer.encode(txt, out_type=str)
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]

# Precision
precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.14
Recall: 0.18
F1-Score: 0.16


#### BPE_2000

In [37]:
pred_tokens = bpe_2000_tokenizer.encode(txt, out_type=str)
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]
# Precision
precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.19
Recall: 0.29
F1-Score: 0.23


#### Mbert_1000

In [44]:
mBtokenizer = mBert()
mB_enc = mBtokenizer.encode_plus(txt, max_length=1000, truncation=True)
pred_tokens = mBertTokenizer.convert_ids_to_tokens(mB_enc['input_ids'])
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]

precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")


Precision: 0.16
Recall: 0.21
F1-Score: 0.18


#### Mbert_2000

In [39]:
mBtokenizer = mBert()
mB_enc = mBtokenizer.encode_plus(txt, max_length=2000, truncation=True)
pred_tokens = mBertTokenizer.convert_ids_to_tokens(mB_enc['input_ids'])
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]

precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.15
Recall: 0.22
F1-Score: 0.18


#### Indi_Bert 1000

In [40]:
indi_teast_token = tokenizer_IndicBert()
indi_test_token = indi_teast_token(txt,max_length=1000, truncation=True)
pred_tokens = indi_teast_token.convert_ids_to_tokens(indi_test_token['input_ids'])
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]


precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.09
Recall: 0.08
F1-Score: 0.09


#### Indi_2000

In [41]:
indi_teast_token = tokenizer_IndicBert()
indi_test_token = indi_teast_token(txt,max_length=2000, truncation=True)
pred_tokens = indi_teast_token.convert_ids_to_tokens(indi_test_token['input_ids'])
pred_tokens = [tok[1:] if tok.startswith('▁') else tok for tok in pred_tokens]


precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.09
Recall: 0.08
F1-Score: 0.09


#### WhiteSpace

In [42]:
strip_data = txt.strip()
if not strip_data:
    pred_tokens=[]
else :
    pred_tokens = txt.split()
precision_value = precision(true_tokens, pred_tokens)
print(f"Precision: {precision_value:.2f}")

# Recall
recall_value = recall(true_tokens, pred_tokens)
print(f"Recall: {recall_value:.2f}")

# F1-score
f1_score_value = f1_score(true_tokens, pred_tokens)
print(f"F1-Score: {f1_score_value:.2f}")

Precision: 0.43
Recall: 0.52
F1-Score: 0.47


## Q6.
### After finding the tokens in Q3 and comparing them with all the tokenizers with different vocab sizes, the results are as follows:
#### order -> unigram_1000 ->unigram_2000 -> bpe_1000 -> bpe_2000 -> mBert_1000 -> mBert_2000 -> Indi_1000 ->Indi_2000 ->whitespace

#### Precision = 0.22 -> 0.24 -> 0.14 -> 0.19 -> 0.16 -> 0.15 -> 0.09 -> 0.09 -> 0.43
#### Recall    = 0.25 -> 0.32 -> 0.18 -> 0.29 -> 0.21 -> 0.22 -> 0.08 -> 0.08 -> 0.52
#### F1        = 0.23 -> 0.27 -> 0.16 -> 0.23 -> 0.18 -> 0.18 -> 0.09 -> 0.09 -> 0.47

#### These models have varying results but have room for improvement. White space tokenizer performs well because the truth value is given by humans as it makes tokens because of whitespace. Indic Bert has the lowest values among all of them.
